### PR TITLE
update binance ratelimit

### PR DESCRIPTION
**Pull request by frosty00**
Got banned testing a trading algo

```
Network related error: binance 429 Too Many Requests {"code":-1003,"msg":"Too many requests; current limit is 1200 requests per minute. Please use the websocket for live updates to avoid polling the API."} <class 'ccxt.base.errors.DDoSProtection'>
Sleeping for 2 seconds...
Network related error: binance 429 Too Many Requests {"code":-1003,"msg":"Too many requests; current limit is 1200 requests per minute. Please use the websocket for live updates to avoid polling the API."} <class 'ccxt.base.errors.DDoSProtection'>
Sleeping for 4 seconds...
Network related error: binance 418 I'm a Teapot {"code":-1003,"msg":"Way too many requests; IP banned until 1519726649240. Please use the websocket for live updates to avoid bans."} <class 'ccxt.base.errors.DDoSProtection'>
Sleeping for 8 seconds...
Network related error: binance 418 I'm a Teapot {"code":-1003,"msg":"Way too many requests; IP banned until 1519726649240. Please use the websocket for live updates to avoid bans."} <class 'ccxt.base.errors.DDoSProtection'>
Sleeping for 16 seconds...
Network related error: binance 418 I'm a Teapot {"code":-1003,"msg":"Way too many requests; IP banned until 1519726649240. Please use the websocket for live updates to avoid bans."} <class 'ccxt.base.errors.DDoSProtection'>
```

Also a troll use of a http error code by an exchange 👀

---

Can you check that 1200 requests per minute means rateLimit = 1200?

### CODE REVIEW

**Comment by frosty00** [comment]
Oh my its more complicated than I thought... 

https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#limits

**Comment by frosty00** [comment]
From what I gather:

```
[{'interval': 'MINUTE', 'limit': 1200, 'rateLimitType': 'REQUESTS'},
 {'interval': 'SECOND', 'limit': 10, 'rateLimitType': 'ORDERS'},
 {'interval': 'DAY', 'limit': 100000, 'rateLimitType': 'ORDERS'}]
```
and 
             {'baseAsset': '123',
              'baseAssetPrecision': 8,
              'filters': [{'filterType': 'PRICE_FILTER',
                           'maxPrice': '100000.00000000',
                           'minPrice': '0.00000100',
                           'tickSize': '0.00000100'},
                          {'filterType': 'LOT_SIZE',
                           'maxQty': '100000.00000000',
                           'minQty': '0.00100000',
                           'stepSize': '0.00100000'},
                          {'filterType': 'MIN_NOTIONAL',
                           'minNotional': '0.00100000'}],
              'icebergAllowed': True,
              'orderTypes': ['LIMIT',
                             'LIMIT_MAKER',
                             'MARKET',
                             'STOP_LOSS_LIMIT',
                             'TAKE_PROFIT_LIMIT'],
              'quoteAsset': '456',
              'quotePrecision': 8,
              'status': 'TRADING',
              'symbol': '123456'}

**Comment by frosty00** [comment]
Could you explain how the tokenbucket works. Essentially what I would like to implement is on https://github.com/ccxt/ccxt/blob/master/js/binance.js#L873 when the exchange returns a 429 error code it means they are about to ban. Therefore I would like to increase the rateLimit or sleep when we recieve this response.

From the docs:

> HTTP 429 return code is used when breaking a request rate limit.
> HTTP 418 return code is used when an IP has been auto-banned for continuing to send requests after receiving 429 codes.

**Comment by mkutny** [comment]
You may want to read a [more detailed description](https://en.wikipedia.org/wiki/Token_bucket) but the basic idea is the following:
- there is a bucket of limited capacity which holds some "usage" tokens
- every time you use something a corresponding amount of tokens is removed from the bucket
- when the bucket holds fewer tokens that are required to use something you have to wait
- every time tick a predefined amount of tokens refills the bucket if it's not full

This basically allows non-linear expenditure.

Right now its max capacity is just 1 second, so for exchanges with `rateLimit` > 1000 ms it's basically linear:
https://github.com/ccxt/ccxt/blob/d6087050f8c0df7ea3931d619201ae352c3eb0f7/python/ccxt/base/exchange.py#L223-L229

But if `rateLimit` is less than 1000 ms, you will be making calls in "bursts" (and this might be an issue if an exchange expects linearity, e.g. expects "1 request per 500 ms" but not "2 requests per second" - see the difference?).

Still, for binance the bucket refills with speed of 2 requests per second which is way below their limits (10 orders/sec). In other words, DDoS errors that you see should not have happened... until their real limit is "1 request per 100 ms". Or there is a bug somewhere. And as no one encountered this issue before I'm more inclined to believe in the latter.

So before you started tweaking the rate limiter I'd be really curious to know how did you manage to get those DDoSes )

**Comment by frosty00** [comment]
Hmmm that is very odd indeed. I'm going to have to test it more thoroughly and get back to you. I'm surprised as well since I was using enable_rate_limit = True. I was doing a lot of order making and cancelling (since it was an outbidding bot).

Okay fast-forward to me having done more testing. It seems that the culprit is `fetch_open_orders`. Now I'm not sure why this is because their API says `privateGetOpenOrders` should have a weight of 1 or 2 and `fetch_orders` should have a weight of 5 - `privateGetAllOrders`. However you can test it yourself and obviously something is wrong in their docs.

The error can be recreated with:

    while True:
        print(binance.fetch_open_orders())

I still think that a better way of approaching this is edit the token bucket when we recieve a 429 response ... pherhaps set capacity to zero?

**Comment by mkutny** [comment]
Can confirm - throws on `fetchOpenOrders()`.

> I still think that a better way of approaching this is edit the token bucket when we recieve a 429 response ... perhaps set capacity to zero?

I think so. At the same time it's a *part* of the solution. Because you don't know how long to wait. The proper way would be assigning different weights to different methods.

**Comment by mkutny** [comment]
> Now I'm not sure why this is because their API says privateGetOpenOrders should have a weight of 1 or 2

Well, nope:

> When all symbols are returned, the number of requests counted against the rate limiter is equal to the number of symbols currently trading on the exchange.

```javascript
console.log(binance.symbols.length) // 270
```

**Comment by frosty00** [comment]
Well the more basic solution is to simply set the symbol in fetch_open_orders. However, I think it using different assigned weights would be a fun challenge also ;)

**Comment by mkutny** [comment]
> However, I think it coding different assigned weights would be a fun challenge also ;)

Absolutely. This would be a solid foundation to switch other exchanges to token buckets too.

BTW, I don't think that the current binance's behavior of sending requests "in packs of two" is intended, so I think it needs to be fixed anyway.

File: js/binance.js [hunk 1]
```diff
@@ -722,8 +721,8 @@ module.exports = class binance extends Exchange {
     }
 
     async fetchOpenOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
-        // if (!symbol)
-        //     throw new ExchangeError (this.id + ' fetchOpenOrders requires a symbol param');
+        if (!symbol)
```
Comment (kroitor) [inline js/binance.js]: Binance says, the symbol isn't mandatory for this call.
https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#current-open-orders-user_data

**Comment by kroitor** [comment]
The most rudimentary type of protection within current codebase would be:
```
// pseudocode
fetch_open_orders (...) {

    // remember the old cost
    old_cost = self.tokenBucket['cost'];

    // set new call cost
    self.tokenBucket['cost'] = 10;

    // do the call
    // ...

    // restore the old cost
    self.tokenBucket['cost'] = old_cost;
}
```

However, a much better way to be extendable and portable would be to start defining call costs in the API schema definition in describe().

**Comment by frosty00** [comment]
@kroitor the reason I added the symbol requirement is to avoid other people getting banned. Calling fetchOpenOrders without one carries a weight of the total number of symbols traded on the exchange divided by two which is ~ 135. Therefore only a couple of calls to fetchOpenOrders can lead to ips getting banned which is going to be confusing for new users using `enableRateLimit = True`. I propose we at least issue a warning of some sort until we have the weight code figured out.

**Comment by kroitor** [comment]
@frosty00 yep, I agree! We can do it in the same way as here: 
- https://github.com/ccxt/ccxt/blob/master/js/okcoinusd.js#L143
- https://github.com/ccxt/ccxt/blob/master/js/okcoinusd.js#L346

**Comment by kroitor** [comment]
@frosty00 I'll add it shortly.

**Pull request merged**